openedx/core/djangoapps/discussions/docs/decisions/0004-in-context-discussions-linking.rst

```diff
@@ -0,0 +1,131 @@
+-------------
+
+Most of these configuration entries would be right at home in the
+`DiscussionsConfiguration` model in `plugin_settings`, however since they need
+to be available during course import-export, they should be stored in the
+course object itself.
+
+There is already a way to associate a discussion ID with a Discussion XBlock
+using its usage key. This same mechanism can be used to associate a Unit usage
+key with a corresponding discussion id.
+
+However the current mechanism has a few issues. It is stored as a JSON
+structure in the `DiscussionsIdMapping` model which has course id and a mapping
+of the discussion id to the xblock usage key in a single dict.
+
+This is OK for the existing setup because this is just a caching mechanism and
+the source of truth for this mapping is the XBlock itself, which stores the
+discussion id. On course publish this information is cached to
+`DiscussionsIdMapping`.
+
+For the new discussions system though, this mapping would be the source of
+truth for the link between discussions and units, so we should use a model
+where each such link is encoded as a row in the database.
+
+Decision
+--------
+
+Since the discussions settings need to be stored in the course structure we
+should create a new JSON structure in the course that matches the structure
+of `plugin_settings`. This can then be used to store not just the settings
+for the inbuilt discussions provider, but for any discussions provider in the
+future.
+
+When a course is published, we can copy over all the `plugin_settings` to the
+course structure with the following structure:
+
+.. code-block:: JSON
+
+    {
+        "edx-next": {
+            "discussions_enable_in_context": bool,
+            "discussions_enable_graded_units": bool,
+            "discussions_group_at_subsection": bool,
+            "discussions_custom_visibility": bool,
+        }
+    }
+
+The `edx-next` key here represents the provider id, allowing for potentially
+multiple provider configs to coexist in case of switching providers etc.
+
+To store Unit-level discussions settings, we can simply add a boolean field
+to the Unit block that specifies whether it is discussable or not. To be
+consistent with the above names we can call this field `discussions_enabled`.
+
+When any of the above settings are changed, the code handling the API call will
+go through all the units and match them against the above criterion, i.e. if
+`discussions_enable_graded_units` is false, and an existing graded unit is
+discussable (whether because an existing unit was switched to gradable, or this
+setting was), then it's `discussions_enabled` flag will be turned off.
+
+As part of the course published event, we will also traverse through all the
+Units in the course that have the flag enabled, and create the corresponding
+discussion topics in `cs_comments_service` and add a mapping. If an existing
+unit with discussions is removed, we will disable the link but not delete the
+data.
```

Comment:
This might be a good place to consider what it would take to decouple this from the modulestore, for other apps that might want to implement discussions. Modulestore access is going to be one of the obstacles to testing and building course apps in general.

For instance, if the pattern for implementing an in-context discussion app is "catch the publish event, call modulestore's get_course() and then traverse the tree looking for the data you need", that ties it tightly to edx-platform and the idiosyncrasies of modulestore. Testing will involve building little courses with ItemFactory, further tying it to edx-platform.

But suppose we decide to create a new `DiscussableContentUpdated` signal, one that might even live in the [openedx-events](https://github.com/eduNEXT/openedx-events/) repo. And say that event contains everything we would normally extract from the modulestore to seed our discussion app data model for this course version–course config, lists of usage keys, titles, etc.

In that scenario, there's something that lives in the Studio process that understands the modulestore and knows how to generate a `DiscussableContentUpdated` event, but then that _event_ is the shared dependency that any discussion app has to implement. That will make testing and development of third party alternatives much simpler, because the hard part that touches modulestore will already be done for them. They (and our own discussions app) can be insulated from where that content data comes from, and all the unnecessary details that they shouldn't have to worry about.

It's almost the same amount of code to write. We'd mainly be shifting it so that instead of the app understanding modulestore, we have something in Studio extract data out of the modulestore and push it into a simplified data model for the app. We do something like this with Learning Sequences where there is Studio code that listens for a course publish and [does the necessary extraction](https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/outlines.py#L325-L387) so that the `learning_sequences` app avoids the modulestore dependency.

Reply:
FYI @mariajgrimaldi and @felipemontoya, as this might be a good use case for openedx-events, but I don't know when that might be stable enough for general use.

Reply:
I think that makes a lot of sense and makes the process a lot cleaner. In fact, since the only thing needed here are the usage keys for units, with some minimal context, we can probably use use the existing blocks or learning sequences APIs to get the relevant data, and the signal can be lighter.

Reply:
Learning Sequences doesn't currently have Unit data (though I do want that later). Since you're hooking into the publish cycle and block transformers are not guaranteed to have run yet, I'd suggest just going straight to the modulestore for generating this event.

Reply:
... and sadly, I think you'll need to do the normal crawling-over-the-course iteration, because using query params on the modulestore fetch to grab you just `vertical` instances will also return you nested `vertical` blocks that are inside of content library blocks, which probably isn't what you want. :-(

Reply:
OK. I will update the ADR to take this into account.

Reply:
Sorry to be late to this discussion. Still I better leave my comments here.

About the decoupling via an openedx-event, yes, that is exactly what the framework supports. According to the comment by @xitij2000 

> In fact, since the only thing needed here are the usage keys for units, with some minimal context

Then we would also want an attr class that can hold the required data using python basic datastructures and usage keys (we already support opaque_keys in the attr classes).

In terms of availability, we are doing the final push to get the first 3 events in the the https://github.com/edx/edx-platform/pull/28266 in the next few days and we expect to get 8 events supported before maple is cut. By then also filters should be available although those are not needed for this I think.

Reply:
@felipemontoya I think that sounds perfect for this use case. 
Good to know it's close to landing in the platform!